common/scala/src/main/scala/whisk/core/containerpool/logging/ElasticSearchRestClient.scala

```diff
@@ -0,0 +1,212 @@
+case class EsQueryRange(key: String, range: EsRange, value: String)
+case class EsQueryBoolMatch(key: String, value: String)
+case class EsQueryOrder(field: String, kind: EsOrder)
+case class EsQuerySize(size: Integer)
+case class EsQueryAll() extends EsQueryMethod
+case class EsQueryMust(matches: Vector[EsQueryBoolMatch], range: Option[EsQueryRange] = None) extends EsQueryMethod
+case class EsQueryMatch(field: String, value: String, matchType: Option[EsMatch] = None) extends EsQueryMethod
+case class EsQueryTerm(key: String, value: String) extends EsQueryMethod
+case class EsQueryString(queryString: String) extends EsQueryMethod
+case class EsQuery(query: EsQueryMethod,
+                   sort: Option[EsQueryOrder] = None,
+                   size: Option[EsQuerySize] = None,
+                   aggs: Option[EsQueryAggs] = None)
+
+// Schema of ES query results
+case class EsSearchHit(source: JsObject)
+case class EsSearchHits(hits: Vector[EsSearchHit])
+case class EsSearchResult(hits: EsSearchHits)
+
+object ElasticSearchJsonProtocol extends DefaultJsonProtocol {
+
+  implicit object EsQueryMatchJsonFormat extends RootJsonFormat[EsQueryMatch] {
+    def read(query: JsValue) = ???
+    def write(query: EsQueryMatch) = {
+      query.matchType match {
+        case Some(matchType) =>
+          JsObject(
+            "match" -> JsObject(
+              query.field -> JsObject("query" -> query.value.toJson, "type" -> matchType.toString.toJson)))
+        case None => JsObject("match" -> JsObject(query.field -> JsObject("query" -> query.value.toJson)))
+      }
+    }
+  }
+
+  implicit object EsQueryTermJsonFormat extends RootJsonFormat[EsQueryTerm] {
+    def read(query: JsValue) = ???
+    def write(query: EsQueryTerm) = JsObject("term" -> JsObject(query.key -> query.value.toJson))
+  }
+
+  implicit object EsQueryStringJsonFormat extends RootJsonFormat[EsQueryString] {
+    def read(query: JsValue) = ???
+    def write(query: EsQueryString) =
+      JsObject("query_string" -> JsObject("query" -> query.queryString.toJson))
+  }
+
+  implicit object EsQueryRangeJsonFormat extends RootJsonFormat[EsQueryRange] {
+    def read(query: JsValue) = ???
+    def write(query: EsQueryRange) =
+      JsObject("range" -> JsObject(query.key -> JsObject(query.range.toString -> query.value.toJson)))
+  }
+
+  implicit object EsQueryBoolMatchJsonFormat extends RootJsonFormat[EsQueryBoolMatch] {
+    def read(query: JsValue) = ???
+    def write(query: EsQueryBoolMatch) = JsObject("match" -> JsObject(query.key -> query.value.toJson))
+  }
+
+  implicit object EsQueryMustJsonFormat extends RootJsonFormat[EsQueryMust] {
+    def read(query: JsValue) = ???
+    def write(query: EsQueryMust) = query.range match {
+      case Some(range) =>
+        JsObject("bool" -> JsObject("must" -> query.matches.toJson, "filter" -> range.toJson))
+      case None =>
+        JsObject("bool" -> JsObject("must" -> query.matches.toJson))
+    }
```

Comment:
Can be simplified to:

```scala
def write(query: EsQueryMust) = {
  val bool = Map("must" -> query.matches.toJson) ++ query.range.map(r => "filter" -> r.toJson)
  JsObject("bool" -> bool)
}
```

It basically only adds the filter to the `Map` if range is `Some` (because then `++` will actually have an element to add). Adding an `Option` to a collection via `++` is generally a great trick to build collections with...well... optional values.